src/shared/openssl-util.c

```diff
@@ -681,28 +644,26 @@ int digest_and_sign(
 int string_hashsum(
                 const char *s,
                 size_t len,
-                const EVP_MD *md_algorithm,
+                const char *md_algorithm,
                 char **ret) {
 
-        uint8_t hash[EVP_MAX_MD_SIZE];
-        size_t hash_size;
-        char *enc;
         int r;
 
-        hash_size = EVP_MD_size(md_algorithm);
-        assert(hash_size > 0);
+        assert(s || len == 0);
+        assert(md_algorithm);
+        assert(ret);
 
-        r = openssl_hash(md_algorithm, s, len, hash, NULL);
+        _cleanup_free_ void *digest = NULL;
+        size_t digest_size;
+        r = openssl_digest(md_algorithm, s, len, &digest, &digest_size);
         if (r < 0)
                 return r;
 
-        enc = hexmem(hash, hash_size);
-        if (!enc)
+        *ret = hexmem(digest, digest_size);
```

Comment:
`ret` should not be overwritten in case of error, so keep the previous temp variable around and assign on success

Reply:
ack